components/calendar/src/any_calendar.rs

```diff
@@ -211,10 +244,44 @@ macro_rules! match_cal {
     };
 }
 
+/// Error returned when comparing two [`Date`]s with [`AnyCalendar`].
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[non_exhaustive]
+pub enum AnyCalendarUntilError {
+    /// The calendars of the two dates being compared are not equal.
+    ///
+    /// To compare dates in different calendars, convert them to the same calendar first.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::calendar::Date;
+    /// use icu::calendar::cal::AnyCalendarUntilError;
+    ///
+    /// let d1 = Date::try_new_gregorian(2000, 1, 1).unwrap().to_any();
+    /// let d2 = Date::try_new_hebrew(5780, 1, 1).unwrap().to_any();
+    ///
+    /// assert!(matches!(
+    ///     d1.until_with_options(&d2, Default::default()),
+    ///     Err(AnyCalendarUntilError::MismatchedCalendars),
+    /// ));
+    ///
+    /// // To compare the dates, convert them to the same calendar,
+    /// // such as ISO.
+    ///
+    /// assert!(matches!(
+    ///     d1.to_iso().until_with_options(&d2.to_iso(), Default::default()),
+    ///     Ok(_)
+    /// ));
+    /// ```
+    MismatchedCalendars,
+}
+
 impl crate::cal::scaffold::UnstableSealed for AnyCalendar {}
 impl Calendar for AnyCalendar {
     type DateInner = AnyDateInner;
     type Year = YearInfo;
+    type UntilError = AnyCalendarUntilError;
```

Comment:
issue: I think this should go on a new trait, not calendar. AnyCalendar is the only calendar where the date difference is fallible, but this makes all calendars fallible. we can impl `Date::since` infallibly for a new `trait CalendarDiff`, and implement it fallibly as `Date::try_since` for `Date<AnyCalendar>`

Reply:
but you still get a `Result` that you have to unwrap, `unwrap_infallible` doesn't exist, and even when it does, is not an ergonomic API.

Reply:
actually the docs currently say "In case the specific calendar objects differ on data, the data for the first calendar is used, and `date2` may be converted if necessary.". Why don't we do that and make the AnyCalendar method infallible?

Reply:
One of these days we'll have `.into_ok()`, and in the mean time, I wrote in the documentation that you can do `let Ok(duration) = date1.until(&date2, options)` and it works since the `Err` variant of the result is uninhabited.

Reply:
I'm open to tweaking the APIs on `Date` as @robertbastian suggested.

If we do that, then should we try for consistency with the `add` functions?

Also, I want to note that I'm not intending to go for ergonomics in this PR. There is a lot of room we have to improve ergonomics, and I consider that very explicitly out of scope. I want to focus on landing the minimal usable API here. I'm open to prefixing functions by `try_*` if it gives us more room to improve in the future.

Reply:
I'm not auto-converting in AnyCalendar any more. I think it's a big footgun.

Reply:
In that case I think I prefer having Result everywhere. `let Ok(..)` is fine. Rust has `Infallible` errors in a lot of places.